compiler/rustc_resolve/src/def_collector.rs

```diff
@@ -300,14 +303,15 @@ impl<'a, 'b, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'b, 'tcx> {
                 // we must create two defs.
                 let closure_def =
                     self.create_def(expr.id, DefPathData::ClosureExpr, DefKind::Closure, expr.span);
-                match closure.asyncness {
-                    Async::Yes { closure_id, .. } => self.create_def(
+                match closure.coro_kind {
+                    Some(CoroutineKind::Async { closure_id, .. }) => self.create_def(
                         closure_id,
                         DefPathData::ClosureExpr,
                         DefKind::Closure,
                         expr.span,
                     ),
-                    Async::No => closure_def,
+                    Some(CoroutineKind::Gen { .. }) => unimplemented!(),
```

Comment:
This could work without implementing gen closures just by matching on `Async | Gen` in the same arm.